### PR TITLE
Shuffle Primitive

### PR DESCRIPTION
**Pull request by parsa**
This PR adds the shuffle primitive to Phylanx, which is the equivalent of `np.random.shuffle`. This primitive was mentioned in #200.

Also:
* Fixes a bug in variable primitive where unwanted copies were made during the assignment of a PhySL variable from PhySL Rvalue (`DynamicMatrix`)
* Updates matrix row & column iterator to make their instances copyable

### CODE REVIEW

File: src/execution_tree/primitives/shuffle_operation.cpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+        return primitive_argument_type(std::move(x));
+    }
+
+    primitive_argument_type shuffle_operation::shuffle_2d(args_type&& args) const
+    {
+        auto x = args[0].matrix();
+        auto x_begin = util::matrix_row_iterator<decltype(x)>(x);
+        auto x_end = util::matrix_row_iterator<decltype(x)>(x, x.rows());
+        std::shuffle(x_begin, x_end, std::mt19937{ std::random_device{}() });
```
Comment (hkaiser) [inline src/execution_tree/primitives/shuffle_operation.cpp]: Would you see a way not to create a new generator and not calling `std::random_device` for each invocation of this primitive?

File: phylanx/util/matrix_iterators.hpp [hunk 2]
```diff
@@ -7,6 +7,7 @@
 #if !defined(PHYLANX_MATRIX_ITERATORS)
 #define PHYLANX_MATRIX_ITERATORS
 
+#include <algorithm>
```
Comment (hkaiser) [inline phylanx/util/matrix_iterators.hpp]: What do you need this #include for?

File: src/execution_tree/primitives/shuffle_operation.cpp [hunk 3]
```diff
@@ -44,6 +44,8 @@ namespace phylanx { namespace execution_tree { namespace primitives
     };
 
     ///////////////////////////////////////////////////////////////////////////
+    std::mt19937 shuffle_operation::rand_machine{std::random_device{}()};
```
Comment (hkaiser) [inline src/execution_tree/primitives/shuffle_operation.cpp]: You're missing `#include <random>` for this file.

File: phylanx/util/matrix_iterators.hpp [hunk 4]
```diff
@@ -1,17 +1,46 @@
...
+    //         caused by matrix_row_iterator not being a real random access
+    //         iterator. Dereferencing matrix_row_iterator does not yield a
+    //         true reference but only a temporary blaze::Row holding true
+    //         references.
+    //
+    // A real fix for this problem is proposed in PR0022R0
+    // (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0022r0.html)
+    //
+    template <typename T>
+    HPX_FORCEINLINE
```
Comment (hkaiser) [inline phylanx/util/matrix_iterators.hpp]: I think this should not be added to the iterators, it's merely a facility that should be visible to the shuffle primitive (i.e. should be placed in shuffle_primitive.cpp).

File: phylanx/util/matrix_iterators.hpp (hunk 2 above)
Comment (parsa) [reply phylanx/util/matrix_iterators.hpp]: I don't need it. Inspect was complaining.

File: src/execution_tree/primitives/shuffle_operation.cpp (hunk 3 above)
Comment (parsa) [reply src/execution_tree/primitives/shuffle_operation.cpp]: It's in the header file.

File: src/execution_tree/primitives/shuffle_operation.cpp (hunk 1 above)
Comment (parsa) [reply src/execution_tree/primitives/shuffle_operation.cpp]: Done.

File: phylanx/util/matrix_iterators.hpp (hunk 4 above)
Comment (parsa) [reply phylanx/util/matrix_iterators.hpp]: I think I got it right. Please check the updated code.

**Pull request merged**